services/bootnode/src/main/java/teku/pegasys/teku/services/bootnode/BootnodeService.java

```diff
@@ -0,0 +1,119 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package teku.pegasys.teku.services.bootnode;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.time.Duration;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+import tech.pegasys.teku.infrastructure.async.AsyncRunner;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.networking.eth2.P2PConfig;
+import tech.pegasys.teku.networking.p2p.discovery.DiscoveryService;
+import tech.pegasys.teku.networking.p2p.discovery.discv5.DiscV5Service;
+import tech.pegasys.teku.networking.p2p.libp2p.LibP2PNetwork.PrivateKeyProvider;
+import tech.pegasys.teku.networking.p2p.libp2p.LibP2PPrivateKeyLoader;
+import tech.pegasys.teku.networking.p2p.network.config.NetworkConfig;
+import tech.pegasys.teku.networks.Eth2NetworkConfiguration;
+import tech.pegasys.teku.service.serviceutils.Service;
+import tech.pegasys.teku.service.serviceutils.ServiceConfig;
+import tech.pegasys.teku.spec.schemas.SchemaDefinitionsSupplier;
+import tech.pegasys.teku.storage.store.FileKeyValueStore;
+import tech.pegasys.teku.storage.store.KeyValueStore;
+
+public class BootnodeService extends Service {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private static final String KEY_VALUE_STORE_SUBDIRECTORY = "kvstore";
+  private static final int PRINT_KNOWN_NODES_COUNT_INTERVAL_IN_SECONDS = 5;
+
+  private final DiscoveryService discoveryService;
+  private final AsyncRunner asyncRunner;
+
+  public BootnodeService(
+      final ServiceConfig serviceConfig,
+      final NetworkConfig networkConfig,
+      final P2PConfig p2PConfig,
+      final Eth2NetworkConfiguration eth2NetworkConfiguration) {
+    final MetricsSystem metricsSystem = serviceConfig.getMetricsSystem();
+    checkNotNull(metricsSystem);
+
+    asyncRunner =
+        serviceConfig.createAsyncRunner(
+            "bootnode",
+            eth2NetworkConfiguration.getAsyncP2pMaxThreads(),
+            eth2NetworkConfiguration.getAsyncP2pMaxQueue());
+    final KeyValueStore<String, Bytes> kvStore =
+        new FileKeyValueStore(
+            serviceConfig
+                .getDataDirLayout()
+                .getBootnodeDataDirectory()
+                .resolve(KEY_VALUE_STORE_SUBDIRECTORY));
+
+    final SchemaDefinitionsSupplier currentSchemaDefinitionsSupplier =
+        () -> eth2NetworkConfiguration.getSpec().getGenesisSchemaDefinitions();
```

Comment:
I wonder if this is ok or if we need to make it a function of the current slot... 🤔

Reply:
Looks like it is safe as down the track it only relies on `SszBitvector` type anyway,